icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -194,6 +194,65 @@ class SimpleUnitIdentifiersSink : public icu::ResourceSink {
     int32_t outIndex;
 };
 
+/**
+ * A ResourceSink that collects information from `unitQuantities` in the `units`
+ * resource (see `units.txt`). unitQuantity values are collected into a char*
+ * array, while unitQuantity keys are added added to a TrieBuilder, with
+ * associated values being the index into the aforementioned char* array.
+ */
+class UnitQuantitiesSink : public icu::ResourceSink {
+  public:
+    /**
+     * Constructor.
+     * @param out Array of char* to which unitQuantity values will be written.
+     * @param outSize The size of the `out` array.
+     * @param trieBuilder The trie builder to which the keys (base units) of
+     *     each unitQuantity will be added, each with value being the offset
+     *     into `out`.
+     */
+    explicit UnitQuantitiesSink(char **out, int32_t &outSize, BytesTrieBuilder &trieBuilder)
+        : outQuantitiesArray(out), outSize(outSize), trieBuilder(trieBuilder), outIndex(0) {}
+
+    void put(const char * /*key*/, ResourceValue &value, UBool /*noFallback*/, UErrorCode &status) {
+        ResourceArray array = value.getArray(status);
+        if (U_FAILURE(status)) return;
+
+        if (outIndex + array.getSize() > outSize) {
+            status = U_INDEX_OUTOFBOUNDS_ERROR;
+            return;
+        }
+
+        for (int32_t i = 0; array.getValue(i, value); ++i) {
+            U_ASSERT(i < array.getSize());
+            U_ASSERT(outIndex < outSize);
+            ResourceTable table = value.getTable(status);
+            U_ASSERT(table.getSize() == 1);
+            const char *key;
+            table.getKeyAndValue(0, key, value);
+            int32_t uTmpLen;
+            // TODO(review): how about we make gQuantities a UChar** instead,
```

Comment:
I do not like having a lot of TODOs in the code, you can have a small issue instead. Or you can review it right now.

Reply:
This TODO is a question to the review - i.e. the intent is indeed to discuss/decide, and fix this, as part of the review.

Reply:
In response to the review question: I strongly prefer if gQuantities is a `UChar**`, such that it contains pointers into the data bundle rather than owning a bunch of strings that might never be used.  I might also look into whether you can change `prefs.getPreferencesFor` to accept a `UChar*` as well; try to move the conversion as close as possible to where it is actually needed.  When you do this, you sometimes discover that you don't actually need to do any conversion at all.

Reply:
Agreed, I've been coming even more to that conclusion with the latest addition (thereby unnecessarily having to repeat the UChar -> char* conversion).

**Not yet done. I'll request another review once I'm done fixing that.**

Reply:
As of 2741a2d I've now got "getUnitQuantity" doing the conversion from UChar* to char*. I also looked at taking it further, having getPreferencesFor accept a UChar*, however it turns out, the "category" comes from a key in the resources, not a value, so is itself a char*. Some conversion thus needs to happen - either I convert all "categories" to UChar at loading time, or I convert all unitQuantities from UChar* to char* at loading time (what I used to have, before 2741a2d), or some conversion happens at lookup time when constructing a NumberFormatter. For now, I'm leaving it with that - conversion happening at getUnitQuantity time.